src/sys/socket/addr.rs

```diff
@@ -1014,4 +1101,227 @@ pub mod sys_control {
             fmt::Display::fmt(self, f)
         }
     }
-}
\ No newline at end of file
+}
+
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[derive(Clone, Copy)]
+pub struct EtherAddr(pub libc::sockaddr_ll);
+
+#[cfg(any(target_os = "dragonfly",
+          target_os = "freebsd",
+          target_os = "ios",
+          target_os = "macos",
+          target_os = "netbsd",
+          target_os = "openbsd"))]
+#[derive(Clone, Copy)]
+pub struct EtherAddr(pub libc::sockaddr_dl);
+
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+impl EtherAddr {
+    pub fn new(sll: libc::sockaddr_ll) -> Option<EtherAddr> {
+        Some(EtherAddr(sll))
+    }
+
+    pub fn is_empty(&self) -> bool {
+        false
+    }
+
+    pub fn addr(&self) -> [u8; 6] {
+        [self.0.sll_addr[0],
+            self.0.sll_addr[1],
+            self.0.sll_addr[2],
+            self.0.sll_addr[3],
+            self.0.sll_addr[4],
+            self.0.sll_addr[5]]
+    }
+}
+
+#[cfg(any(target_os = "dragonfly",
+          target_os = "freebsd",
+          target_os = "ios",
+          target_os = "macos",
+          target_os = "netbsd",
+          target_os = "openbsd"))]
+impl EtherAddr {
+    pub fn new(sdl: libc::sockaddr_dl) -> Option<EtherAddr> {
+        let sdl_wrap = EtherAddr(sdl);
+        match sdl_wrap.is_empty() {
```

Comment:
Can this ever happen in practice? Would it be a system error if so? Might be better to implement with an `assert!` or a `panic!` if so.